Atomic.md

```diff
@@ -0,0 +1,278 @@
+
+        -   A incomplete placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   A incomplete placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .temp yet
+
+    -   The rename() to copy the completed placeholder from .temp into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+With this approach, each placeholder file that will be created will be moved in one at a time to their final location rather than moving all placeholders that may be present in a folder at once. 
+
+In a case where we crash (or the machine crashes, or the user kills the mount process) while expanding a placeholder directory, we will be able to recover by replacing all placeholders that are present in that folder and continuing the enumeration. We will not try to short circuit this process and restart the enumeration from where we crashed. Any user content that is present in this directory will not be overwritten by the directory enumeration.
+
+Designs where we moved an entire folder's worth of placeholders were considered, but were tabled in favor of this approach (making any expansion/hydration retriable).
+
+All of the potential holes identified above are filled by this design,
+so I believe this to be our best bet to create placeholders in an atomic
+matter.
+
+Miscellanea/Things to Validate
+------------------------------
+
+-   Is it possible to hydrate a placeholder in a directory that isn't
+    fully enumerated? It shouldn't be, will revalidate experimentally.
+
+-   Placeholders that are being created in .temp should be given a
+    temporary name (we'll rename() into the actual name). If we give
+    them the exact name of the final file, there is a hole where some
+    issue occurs, they get abandoned in the .temp directory. Once that
+    happens, we'll encounter issues on a retry where we try creating a
+    file where one exists with the same name.
+
+-   A directory that we're filling with placeholders that has files not
+    owned by us (no bits set) should not have changes made to the files
+    that we don't own.
+
+-   A user could write into one of our unhydrated placeholders (if our
+    kext isn't loaded, we can't block their access to the file). We
+    should validate that we never lose any changes that a user makes in
+    this way.
+
+-   We should clear the .temp directory on mounting to prevent abandon
+    placeholders from cluttering up disk space.
+
+-   All of the notification work will need to be validated under the new
```

Comment:
I think you can just rely on functional tests to handle this. This is sort of assumed for every new feature - make sure you don't break any other feature. Not needed to call it out here, I think.